packages/zoe/contracts/simpleExchange.js

```diff
@@ -22,63 +15,53 @@ import {
 // support partial fills of orders.
 
 export const makeContract = harden((zoe, terms) => {
-  const sellOfferHandles = [];
-  const buyOfferHandles = [];
+  const sellInviteHandles = [];
+  const buyInviteHandles = [];
+  const { assays } = terms;
+  const {
+    rejectOffer,
+    hasValidPayoutRules,
+    swap,
+    areAssetsEqualAtIndex,
+  } = makeHelpers(zoe, assays);
 
-  const simpleExchange = harden({
-    addOrder: async escrowReceipt => {
-      const {
-        offerHandle,
-        offerRules: { payoutRules },
-      } = await zoe.burnEscrowReceipt(escrowReceipt);
-
-      // Is it a valid sell offer?
-      const sellOfferKinds = ['offerAtMost', 'wantAtLeast'];
-      if (hasValidPayoutRules(sellOfferKinds, terms.assays, payoutRules)) {
-        // Save the valid offer
-        sellOfferHandles.push(offerHandle);
-
-        // Try to match
-        const {
-          offerHandles: activeBuyHandles,
-          payoutRulesArray: activeBuyPayoutRules,
-        } = getActivePayoutRules(zoe, buyOfferHandles);
-        for (let i = 0; i < activeBuyHandles.length; i += 1) {
-          if (isMatchingLimitOrder(zoe, payoutRules, activeBuyPayoutRules[i])) {
-            return reallocate(zoe, offerHandle, activeBuyHandles[i]);
+  const makeInvite = () => {
+    const seat = harden({
+      addOrder: () => {
+        // Is it a valid sell offer?
+        if (hasValidPayoutRules(['offerAtMost', 'wantAtLeast'], inviteHandle)) {
+          // Save the valid offer and try to match
+          sellInviteHandles.push(inviteHandle);
+          for (const buyHandle of zoe.getOfferStatuses(buyInviteHandles)
+            .active) {
+            if (areAssetsEqualAtIndex(0, inviteHandle, buyHandle)) {
```

Comment:
Where are you testing `canTradeWith`?

Reply:
inside swap :)

Reply:
If the `canTradeWith` inside swap says false, `swap` throws. This is what `swap` should do, but is not the `canTradeWith` test you want in this loop. As written, for the first where the goods are equal, not `canTrade`, you'll exit the loop with a thrown error.

Reply:
yeah, good point.